src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -1114,7 +1114,8 @@ protected void ExecuteIndexing()
                                 }
                                 catch (IndexWriteException iwe)
                                 {
-                                    HandleWriteErrors(scope, iwe);
+                                    if(DocumentDatabase.DatabaseShutdown.IsCancellationRequested == false)
```

Comment:
They way to handle that would be to not wrap OperationCanceledException with IndexWriteException. From my brief look that this only can happen for OutputMapToReduce because we are using TxMerger there (2 places).

```
                            try
                            {
                                enqueue.GetAwaiter().GetResult();
                            }
                            catch (OperationCanceledException)
                            {
                                throw;
                            }
                            catch (Exception e)
                            {
                                throw new IndexWriteException("", e);
                            }
```

The other advantage of doing that is that index will not continue working (it will close down the indexing thread immediately - in your case you will write indexing errors and other stats - which is not what we probably want, etc)

Also please add Debug.Assert to:
```
                                catch (OperationCanceledException)
                                {
                                    Debug.Assert(_indexingProcessCancellationTokenSource.IsCancellationRequested, "msg");

                                    // We are here only in the case of indexing process cancellation.
                                    scope.RecordMapCompletedReason("Operation canceled.");
                                    return;
                                }
```

Reply:
There can be index error also for `ObjectDisposedException` when the `OutputReduceIndexWriteOperation` try to add operation to disposed `TransactionOperationsMerger`

Reply:
Indexing should be stopped way before TxMerger is disposed.

Reply:
To get exception from awating to `CommitOutputReduceToCollection` here - https://github.com/ravendb/ravendb/blob/d22e2b6c20e5e989f9eba4fe434047a5864cd911/src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceIndexWriteOperation.cs#L43
there are two options:
1. The last `OutputReduceToCollectionCommand` in progress canceled - this will throw `OperationCanceledException`.
2. When trying to enque a new `OutputReduceToCollectionCommand` - this for now throw `ObjectDisposedException`

The `TransactionOperationsMerher.Enque` throw `ObjectDisposedException` when the `_concurrentOperations` is zero.
In the `TransactionOperationsMerher.Dispose` we first decrement `_concurrentOperations` to zero and only then iterate all the running operations in the `TransactionOperationsMerher` and canceling them.

I can add in `CommitOutputReduceToCollection` a check before enqueuing a new `OutputReduceToCollectionCommand` if the `ServerShutdown` canceled but there is still an option the code will pass the check and then `_concurrentOperations` will decrement to zero.

Another option is to throw 'TaskCanceledException` from `TransactionOperationsMerher.Enque`.

Reply:
It's impossible to get the `ObjectDisposed` exception since we are waiting for the indexing thread to finish before we dispose the `TxMerger`.
https://github.com/ravendb/ravendb/blob/v5.1/src/Raven.Server/Documents/Indexes/Index.cs#L298
https://github.com/ravendb/ravendb/blob/v5.1/src/Raven.Server/Documents/Indexes/Index.cs#L309

Reply:
Actually, when I tried to reproduce it I found it more difficult to get `TaskCanceledException`.
Just create an index with enough documents for multiple batches of output reduce to collection.
Put breakpoint here https://github.com/ravendb/ravendb/blob/d22e2b6c20e5e989f9eba4fe434047a5864cd911/src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceIndexWriteOperation.cs#L43
and when it got there just press 'q' and 'y' in the console to shut down the server, continue the debugger and you will get `ObjectDisposedException`.

Reply:
I added the RavenDB dump file to reproduce if you like https://issues.hibernatingrhinos.com/issue/RavenDB-15683#focus=Comments-67-55584.0-0

Reply:
The main point here is the `ObjectDisposedException` is thrown base on `_concurrentOperations` and we use it to block enqueuing more operations to `TransactionOperationsMerher`.

Reply:
Interesting. Can you share the actual stacktrace of `ObjectDisposedExecption`? Maybe we have invalid order of disposals

Reply:
```
Raven.Server.dll!Raven.Server.Documents.TransactionOperationsMerger.ThrowTxMergerWasDisposed() Line 150
Raven.Server.dll!Raven.Server.Documents.TransactionOperationsMerger.Enqueue(Raven.Server.Documents.TransactionOperationsMerger.MergedTransactionCommand cmd) Line 129
Raven.Server.dll!Raven.Server.Documents.Indexes.MapReduce.OutputToCollection.OutputReduceIndexWriteOperation.CommitOutputReduceToCollection() Line 59	
```

Reply:
It looks like the order of disposal is invalid. We first are disposing the merger, then subsystems that are actually using it. From my brief look we should dispose merger last (just before _databaseShutdown.Dispose)

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-16243